userdata/userdata_cli_interop_test.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 Steffen Vogel.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package userdata_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"os/exec"
+	"testing"
+
+	"github.com/google/nftables"
+	"github.com/google/nftables/userdata"
+)
+
+type NftCliMetainfo struct {
+	Version           string `json:"version,omitempty"`
+	ReleaseName       string `json:"release_name,omitempty"`
+	JSONSchemaVersion int    `json:"json_schema_version,omitempty"`
+}
+
+type NftCliTable struct {
+	Family string `json:"family,omitempty"`
+	Name   string `json:"name,omitempty"`
+	Handle int    `json:"handle,omitempty"`
+}
+
+type NftCliChain struct {
+	Family string `json:"family,omitempty"`
+	Table  string `json:"table,omitempty"`
+	Name   string `json:"name,omitempty"`
+	Handle int    `json:"handle,omitempty"`
+}
+
+type NftCliExpr struct{}
+
+type NftCliRule struct {
+	Family  string       `json:"family,omitempty"`
+	Table   string       `json:"table,omitempty"`
+	Chain   string       `json:"chain,omitempty"`
+	Handle  int          `json:"handle,omitempty"`
+	Comment string       `json:"comment,omitempty"`
+	Expr    []NftCliExpr `json:"expr"`
+}
+
+type NftCommand struct {
+	Ruleset interface{}  `json:"ruleset"`
+	Table   *NftCliTable `json:"table,omitempty"`
+	Chain   *NftCliChain `json:"chain,omitempty"`
+	Rule    *NftCliRule  `json:"rule,omitempty"`
+}
+
+type NftCliObject struct {
+	Metainfo *NftCliMetainfo `json:"metainfo,omitempty"`
+	Table    *NftCliTable    `json:"table,omitempty"`
+	Chain    *NftCliChain    `json:"chain,omitempty"`
+	Rule     *NftCliRule     `json:"rule,omitempty"`
+	Add      *NftCommand     `json:"add,omitempty"`
+	Flush    *NftCommand     `json:"flush,omitempty"`
+}
+
+type NftCli struct {
+	Nftables []NftCliObject `json:"nftables"`
+}
+
+func TestCommentInteropGo2Cli(t *testing.T) {
+	wantComment := "my comment"
+
+	c, err := nftables.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c.FlushRuleset()
+
+	table := c.AddTable(&nftables.Table{
+		Name:   "userdata-table",
+		Family: nftables.TableFamilyIPv4,
+	})
+
+	chain := c.AddChain(&nftables.Chain{
+		Name:  "userdata-chain",
+		Table: table,
+	})
+
+	c.AddRule(&nftables.Rule{
+		Table:    table,
+		Chain:    chain,
+		UserData: userdata.PutString(nil, userdata.TypeComment, wantComment),
+	})
+
+	if err := c.Flush(); err != nil {
+		t.Fatal(err)
+	}
+
+	out := bytes.NewBuffer(nil)
+	d := exec.Command("nft", "-j", "list", "table", "userdata-table")
```

Comment:
Tests that modify system state should be guarded behind a flag, like enableSysTests in nftables_test.go.

Even better: make these tests use a throwaway network namespace so that they are safe to run without affecting the rest of the system.

Reply:
Done. To reuse the code between `nftables_test.go` and `userdata_cli_interop_test.go`, I moved the setup/teardown of the NFT connection to `internal/nftest`.